### DISCUSSION TITLE
ESP32_W5500 with multiple SPI slaves

### DISCUSSION THREAD

**Discussion by Avengm** [khoih-prog/AsyncMQTT_Generic#27] [Q&A]
Thank @khoih-prog for this wonderful library. I am trying to integrate multiple spi sensors such as RC522, LCD. with FullyFeatured_ESP32_W5500 example but the esp is consistently rebooting, I guess due to when W5500 slave is being selected at that time other spi is also in use. Is there any way to use spi slaves and w5500 together?

**Comment by khoih-prog** [comment]
> I guess due to when W5500 slave is being selected at that time other spi is also in use. 

Not true, **unless your code / hardware is buggy**. SPI can be used for single-master (ESP32) and multiple-slaves (W5500, etc.)
Certainly with correctly written codes and hardware.

You can isolate and eliminate the culprit(s) by removing them one-by-one

One example is the bad/buggy SD-card hardware design which **doesn't tri-state the SPI MISO even when CS/SS not selected.**

> Is there any way to use spi slaves and w5500 together?

Just careful test and select the correct hardware / libraries

Post the question on Arduino Forum to get advice and help from many experts there. **Certainly with all the info about hardware / libraries / your code.**

**Comment by Avengm** [comment]
Hi @khoih-prog I just shifted to `AsyncWebServer_ESP32_W5500` and `AsyncMQTT_ESP32` libraries. When I comment out the ethernet rest of the spi slaves are working fine. But when it's not commented I get the following error
```
assert failed: spi_hal_setup_trans IDF/components/hal/spi_hal_iram.c:68 (spi_ll_get_running_cmd(hw) == 0)


Backtrace: 0x40083bd9:0x3ffbb7e0 0x4008bd05:0x3ffbb800 0x400921d5:0x3ffbb820 0x40090575:0x3ffbb950 0x400ed668:0x3ffbb990 0x400edc9a:0x3ffbb9e0 0x400edd7c:0x3ffbba10 0x400dcef3:0x3ffbba30 0x400dd009:0x3ffbba90 0x4016a629:0x3ffbbac0 0x400dc86f:0x3ffbbae0 0x4016a613:0x3ffbbb20 0x40117013:0x3ffbbb40

  #0  0x40083bd9:0x3ffbb7e0 in panic_abort at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/panic.c:402
  #1  0x4008bd05:0x3ffbb800 in esp_system_abort at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/esp_system.c:128
  #2  0x400921d5:0x3ffbb820 in __assert_func at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/newlib/assert.c:85
  #3  0x40090575:0x3ffbb950 in spi_ll_master_set_line_mode at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/hal/esp32/include/hal/spi_ll.h:452
      (inlined by) spi_hal_setup_trans at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/hal/spi_hal_iram.c:70
  #4  0x400ed668:0x3ffbb990 in spi_new_trans at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/driver/spi_master.c:561
  #5  0x400edc9a:0x3ffbb9e0 in spi_device_polling_start at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/driver/spi_master.c:958
      (inlined by) spi_device_polling_start at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/driver/spi_master.c:927
  #6  0x400edd7c:0x3ffbba10 in spi_device_polling_transmit at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/driver/spi_master.c:1000
  #7  0x400dcef3:0x3ffbba30 in w5500_read at .pio/libdeps/esp32dev/WebServer_ESP32_W5500/src/w5500/esp_eth/esp_eth_mac_w5500.c:142
  #8  0x400dd009:0x3ffbba90 in emac_w5500_read_phy_reg at .pio/libdeps/esp32dev/WebServer_ESP32_W5500/src/w5500/esp_eth/esp_eth_mac_w5500.c:543 (discriminator 2)
  #9  0x4016a629:0x3ffbbac0 in eth_phy_reg_read at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_eth/src/esp_eth.c:75
      (inlined by) eth_phy_reg_read at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_eth/src/esp_eth.c:66
  #10 0x400dc86f:0x3ffbbae0 in w5500_update_link_duplex_speed at .pio/libdeps/esp32dev/WebServer_ESP32_W5500/src/w5500/esp_eth/esp_eth_phy_w5500.c:99
      (inlined by) w5500_get_link at .pio/libdeps/esp32dev/WebServer_ESP32_W5500/src/w5500/esp_eth/esp_eth_phy_w5500.c:166
  #11 0x4016a613:0x3ffbbb20 in eth_check_link_timer_cb at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_eth/src/esp_eth.c:161
  #12 0x40117013:0x3ffbbb40 in timer_process_alarm at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_timer/src/esp_timer.c:360
      (inlined by) timer_task at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_timer/src/esp_timer.c:386




ELF file SHA256: 17a8dea4a030dbdc

Rebooting...
ets Jul 29 2019 12:21:46
```

**Comment by khoih-prog** [comment]
You have to try by using baby-steps to isolate the problems, especially when having complex issues

Try a single library first (without other SPI hardware) to be sure your W5500 is OK (hardware, wiring, etc.)

1. `WebServer_ESP32_W5500`
2. `AsyncWebServer_ESP32_W5500`

Remember to use `INT` pin or not working at all.

---

Another point is you can move the W5500 or other hardware's SPI to another SPI bus to avoid conflict.

I'm sorry that's all the time I can spend here. I suggest you post for help on Arduino Forum, etc.

**Reply by Avengm** [reply]
Thank you for your suggestion, When I connect W5500 to another SPI hardware everything is working without any conflict. I think since I'm running async server and when any spi slave other than W5500 is active and any W5500 related operation is called then due to multiple active spi slaves esp is rebooting. If I'm right, I was wondering whether AsyncMQTT_ESP32  could detect whether any other spi slave is still in use or not, and If yes it'd postpone the operation momentarily till other spi slaves release the slave line.

**Reply by khoih-prog** [reply]
> I'm running async server and when any spi slave other than W5500 is active and any W5500 related operation is called then due to multiple active spi slaves esp is rebooting.

This is not how SPI bus and SPI software / library are designed, certainly they must rely on the correct hardware / wiring / configuration (**CS/SS, tri-state,** etc.)

Only you can figure out what's wrong with your hardware / code.

I don't believe this is the bug of  the **ESP32 SPI library**

**Reply by Avengm** [reply]
Then I'm assuming that probably INT pin of my hardware seems to be not working, It's connected to GPIO 21 hence `#define INT_GPIO            21` in defines.h. May I please know the reason, "Why INT of W5500 is important?"

**Comment by Avengm** [comment]
Hi @khoih-prog Could you please post an example for the W5500 async web server where W5500 and any other SPI device are connected on the same SPI bus?
